proposals/0043-prefetch.md

```diff
@@ -0,0 +1,142 @@
+
+<!-- prefetch on tap (touchstart, mousedown events) -->
+<a href="/foo" data-astro-prefetch="tap">foo</a>
+
+<!-- prefetch on hover (focusin, focusout, mouseenter, mouseleave events) -->
+<a href="/foo" data-astro-prefetch="hover">foo</a>
+
+<!-- prefetch on enter viewport (IntersectionObserver) -->
+<a href="/foo" data-astro-prefetch="viewport">foo</a>
+
+<!-- disable prefetch if prefetch all by default -->
+<a href="/foo" data-astro-prefetch="false">foo</a>
+```
+You can also configure options in `astro.config.js`:
+
+```js
+prefetch: {
+  // Whether to prefetch all links by default. same as adding `data-astro-prefetch` to all links.
+  // Default false, but true if view transitions is used. User can explicitly set this to false
+  // to disable for view transitions. (I tried coming up a better name)
+  prefetchAll: true,
+  // Default prefetch strategy for `data-astro-prefetch` (no value specified).
+  defaultStrategy: 'hover' // accepts 'tap' and 'viewport' too
+}
+```
+
+Different prefetch strategies have different behaviours and opinions:
+- `tap`: Calls `fetch()` on `touchstart` or `mousedown` events (they are called before `click` event)
+- `hover`: Calls `fetch()` on `focusin`+`focusout` or `mouseenter`+`mouseleave` events. Hover detection kicks in after 80ms delay.
+- `viewport`: Creates a `<link rel="prefetch">` on enter viewport. It has lower priority than `fetch()` to not clog up the request. Intersection detection kicks in after 300ms.
+
+Notes: `hover` and `viewport` only works on `<a />` tags on initial page load (and view transition page load) due to limitations of the events. Unless we watch the entire DOM with MutationObserver but it's not performant.
+
+---
+
+For programmatic usage:
+
+```js
+import { prefetch } from 'astro:prefetch'
+
+prefetch('http://...', { with: 'link' }) // second parameter optional, can specify link/fetch for prefetch priority
+```
+
+# Background & Motivation
+
+With the introduction of View Transitions, it includes partial prefetching code for snappy navigation between pages. We can take this opportunity to support prefetching in core, and share the prefetch behaviour with View Transitions. 
+
+I've started an implementation before an RFC as the initial plan was to simply move `@astrojs/prefetch` to core. However, it would also be a good time to polish up and extend the API.
+
+# Goals
+
+- An option to enable prefetching
+- Enable prefetching via an attribute/hint
+- Enable prefetching for all links by default (required by View transitions)
+- Disable prefetching if all links are enabled by default
+- Different prefetching strategies (click, hover, viewport, etc)
+- Only add JS if using prefetching
+
+# Non-Goals
+
+- Prefetch cache invalidation (Browser relies on cache control)
+- Prefetch external links
+
+# Detailed Design
+
+A prefetch script for client-side is required. It should only be included if `prefetch` is truthy. The script can be injected through the `injectScript` integration API.
```

Comment:
> It should only be included if `prefetch` is truthy.

Do you mean the config option? I think this isn't true if so, it can also be included if someone uses the programmatic API, right?

Reply:
Ah yup. Will clarify that. Now that you mention about the programmatic API, the user would have to enable the `prefetch` config first before they can import it.

I've not covered this error handling if not, so I should also fix that in the implementation PR.

Reply:
Actually, ignore the latter 😅  You can import `prefetch` from `astro:prefetch` without the `prefetch: true` config with the current implementation I think, since it's standalone manual prefetch and will not setup any listeners.  But I think it's better to only allow importing `astro:prefetch` if `prefetch: true` is enabled so it's consistent with our other APIs before.